docs: new variants per environment (#2590)
Unleash/unleash: website/docs/reference/api/legacy/unleash/admin/features-v2.md

    "name": "DemoNew",
    "project": "default",
    "stale": false,
    "type": "release",
    "variants": [
        {
            "name": "blue",
            "overrides": [],
            "stickiness": "default",
            "weight": 1000,
            "weightType": "variable"
        }

>     ]

Reviewer: Were they previously included or was the doc wrong here?

>     ]

Author: I went back in time and this was present up until Unleash 4.17. With 4.18 it got removed, probably because variants were no longer at the feature level. Unfortunately, a breaking change.

>     ]

Author: I left it untouched as it's legacy documentation

## Feature Variants

**Important notes:** 
1. Variants were previously linked to feature toggles making it impossible to have different variants per environment. We are migrating to variants per environment which will allow you to manage variants independently in each environment. 
2. As part of the migration, existing variants will be copied to all the environments of its feature which will respect the behaviour prior to having variants per environment.
3. Clients and SDKs are not affected by this changes as they always access features within one environment.
4. The export format will change to have variants linked to environments. Old exported formats will be gracefully handled if used to import data.

> 5. Because this change adds more data, it won't be possible to down migrate without data loss (i.e. if variants are different per environment, there's no way of representing that in the model where variants were linked to features and applying to all environments)

Reviewer: This is great information, but feels like more of a blog post than reference documentation, I think. We could rewrite it to a section about changes from 4.x to 4.x+1 or 5 or whatever. What do you think about that?

Author: Yes, I agree, it could be part of the release notes or something :thinking:

### Put variants for Feature Toggle {#update-variants}

> **Warning** this affects all environments at once. It's recommended to use [#update-variants-per-environment](#update-variants-per-environment)

Reviewer: We should use [docusaurus' admonitions](https://docusaurus.io/docs/markdown-features/admonitions) when making call-outs in the text.

```suggestion
:::caution 

This endpoint affects all environments at once. If you only want to update a single environment, use [#update-variants-per-environment](#update-variants-per-environment) instead.

:::
```

## Feature Variants

### Put variants for Feature Toggle {#update-variants}

> :::caution 

Reviewer: Formatting again.

```suggestion

:::caution 
```

> From 4.21 variants are tied to an environment. Check our Open API docs for the up-to-date docs: https://docs.getunleash.io/reference/api/unleash/features

Reviewer: Same question here regarding the link? Why are we linking them there?

```suggestion
Since Unleash, 4.21 variants are tied to environments.
```

Reviewer: Got it. So as above: it should be linking the the openapi doc for this specific endpoint, right? `/docs/reference/api/unleash/overwrite-feature-variants.api.mdx`? I'll make a suggestion for that.

Author: yes, I think that'd be the best to start making it our source of truth

> This endpoint affects all environments at once. If you only want to update a single environment, use [#update-variants-per-environment](#update-variants-per-environment) instead.

Reviewer: Where is that link supposed to go? I can't seem to find that ID anywhere in this document? And why is the link text just the anchor?

```suggestion
This endpoint affects all environments at once. If you only want to update a single environment, use [#update-variants-per-environment](#update-variants-per-environment) instead.
```

Author: It was pointing to the section I added in the previous version. Now it should be pointing to the OpenAPI spec maybe

```

### PATCH variants for a feature toggle

> **Warning** this affects all environments at once. It's recommended to use [#update-variants-per-environment](#update-variants-per-environment)

Reviewer: ```suggestion
:::caution

 This endpoint affects all environments at once. If you only want to update a single environment, use [#update-variants-per-environment](#update-variants-per-environment) instead.
```

```

### PATCH variants for a feature toggle

> :::caution

Reviewer: This is just for formatting reasons.

```suggestion

:::caution
```

> From 4.21 variants are tied to an environment. Check our Open API docs for the up-to-date docs: https://docs.getunleash.io/reference/api/unleash/features

Reviewer: I'm not sure what the link does here? What's the purpose of it? Why are you linking them to that page exactly?

```suggestion
Since Unleash 4.21, variants are tied to environments.
```

Reviewer: Sorry, I'm not sure I understand. Why are you linking them there? Are you saying that the OpenAPI docs you're pointing to are the more updated versions of the doc for this specific endpoint?

Author: Isn't this documentation "legacy"? We've found some payloads which are not up-to-date (due to a non-intentional change), but OpenApi should have the correct shapes for these API endpoints. This would avoid maintaining the same documentation in two places... 

But we can talk about it, I'm sensing there's something I'm missing here

Reviewer: Yes, it is. I don't think you're missing anything. This is legacy documentation and shouldn't really be updated (though I like the idea of adding links to the new docs). So yes, this is a good change. I just wanted to understand where you want to link them, but I think I got it! 😄

 This endpoint affects all environments at once. If you only want to update a single environment, use [#update-variants-per-environment](#update-variants-per-environment) instead.

:::

> <ApiRequest verb="patch" url="api/admin/projects/:projectId/features/:featureName/variants" title="Patch variants for a feature toggle trying to apply the changes in all environments (example data)" payload={[{"op": "add", "path": "/1", "value": {

Reviewer: Could we simplify it to this?

```suggestion
<ApiRequest verb="patch" url="api/admin/projects/:projectId/features/:featureName/variants" title="Patch variants for a feature toggle in all environments (example data)" payload={[{"op": "add", "path": "/1", "value": {
```

### PATCH variants for a feature toggle

:::caution

> From 4.21 variants are tied to an environment. Check our Open API docs for the up-to-date docs: https://docs.getunleash.io/reference/api/unleash/features
> 
>  This endpoint affects all environments at once. If you only want to update a single environment, use [#update-variants-per-environment](#update-variants-per-environment) instead.

Reviewer: So something like this? 

```suggestion
This API documentation is no longer maintained. You should use the [OpenAPI docs for this endpoint](/docs/reference/api/unleash/overwrite-feature-variants.api.mdx) instead.

This endpoint affects all environments at once. If you only want to update a single environment, use the [operation for updating variants per enviroment](/docs/reference/api/unleash/overwrite-feature-variants-on-environments.api.mdx) instead.
```

Author: Yes, but I can't find the linked doc, maybe it's autogenerated? You may know better than I do, so feel free to commit this suggestion because despite the link being correct is what I think we should say

Reviewer: Yeah, the doc is generated, but you can generate it yourself. Anyway, this should link to the right place.

}
```

### Put variants for Feature Toggle on environment {#update-variants-per-environment}

<ApiRequest verb="put" url="api/admin/projects/:projectId/features/:featureName/environments/:environment/variants" title="Create (overwrite) variants for a feature toggle under the provided environment (example data)" payload={[
	{
		"name": "variant1",
		"weightType": "fix",
		"weight": 650,
        "payload": {
          "type": "json",
          "value": "{\"key1\": \"value\", \"key2\": 123}"
        },
        "stickiness": "userId",
        "overrides": [ {
          "contextName": "userId",
          "values": ["1", "23"]
        } ]
	},
	{
		"name": "variant2",
		"weightType": "variable",
		"weight": 123
	}
]}/>

This overwrites the current variants for the feature toggle specified in the :featureName parameter for the :environment parameter.
The backend will validate the input for the following invariants

* If there are variants, there needs to be at least one variant with `weightType: variable`

> * The sum of the weights of variants with `weightType: fix` must be below 1000 (< 1000)

Reviewer: Is this correct? It's < and not <= 1000, right?

```suggestion
* The sum of the weights of variants with `weightType: fix` must be strictly less than 1000 (< 1000)
```

Author: Actually, we require to have at least one variable variant, and the sum of the fixed ones can be 1000 (leaving the variable variant at zero). So we should change the doc, but at the same time I'm thinking that putting this behaviour here might lead to inconsistencies if we change it later and makes me wonder about the level of detail we want to have because these nuances could be in the OpenAPI doc

Reviewer: Got it, so it should be something like this instead, then:

```suggestion
* The sum of the weights of variants with `weightType: fix` must be less than or equal to 1000 (≤ 1000)
```

> So we should change the doc, but at the same time I'm thinking that putting this behaviour here might lead to inconsistencies if we change it later and makes me wonder about the level of detail we want to have because these nuances could be in the OpenAPI doc

I'm not entirely sure what you're saying here; could you explain a bit? Anyway, I think this should absolutely go in the OpenAPI doc. This is behavior that people using the endpoints will want to know about. And if it's something that we do end up changing later, then we'll need to update the API docs too.

Author: So, the thing is to add a fixed variant, you first need to add a variable variant. This variable variant needs to have at least weight of 1. So the fixed variant can weight at most 999, but never 1000.

Author: We could say here something like:
> The sum of the weights of variants with `weightType: fix` must be less than or equal to 1000 (which should include at least one variant with `weightType: variable`)

> The backend will also distribute remaining weight up to 1000 after adding the variants with `weightType: fix` together amongst the variants of `weightType: variable`

Reviewer: ```suggestion
The back end will also distribute remaining weight up to 1000 after adding the variants with `weightType: fix` together amongst the variants of `weightType: variable`
```
